File: src/main/java/dev/slne/discord/listener/interaction/button/DiscordButtonListener.java
package dev.slne.discord.listener.interaction.button;

import dev.slne.discord.spring.annotation.DiscordListener;
import dev.slne.discord.spring.annotation.processor.DiscordButtonProcessor;
import dev.slne.discord.spring.annotation.processor.DiscordButtonProcessor.DiscordButtonHandlerHolder;
import javax.annotation.Nonnull;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;

/**
 * The type Discord button listener.
 */
@DiscordListener
public class DiscordButtonListener extends ListenerAdapter {

  private final DiscordButtonProcessor discordButtonProcessor;

  public DiscordButtonListener(DiscordButtonProcessor discordButtonProcessor) {
    this.discordButtonProcessor = discordButtonProcessor;
  }

  @Override
  public void onButtonInteraction(@Nonnull ButtonInteractionEvent event) {
    final DiscordButtonHandlerHolder holder = discordButtonProcessor.getHandler(
        event.getButton().getId());

    if (holder != null) {
      holder.handler().onClick(event);
    }
  }
}

File: src/main/kotlin/dev/slne/discord/listener/interaction/button/DiscordButtonListener.kt
package dev.slne.discord.listener.interaction.button

import dev.minn.jda.ktx.events.listener
import dev.slne.discord.discord.interaction.button.DiscordButtonProcessor
import jakarta.annotation.PostConstruct
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import org.springframework.stereotype.Component

@Component
class DiscordButtonListener(private val jda: JDA, private val processor: DiscordButtonProcessor) {

    @PostConstruct
    fun registerListener() {
        jda.listener<ButtonInteractionEvent> { event ->
            val id = event.button.customId ?: return@listener
            processor.getHandler(id)?.second?.apply { event.onClick() }
        }
    }
}
